### ISSUE TITLE
More than 1000 IoT-Hub devices

### ISSUE THREAD

**Issue by PieterID** [Azure/azure-iot-sdk-csharp#88]
Hi,

I have installed Device Explorer from build 20170413

We have connected more than 1000 devices to our IoT-Hub but only the last 1000 are showed in the explorer view. 

I checked the source code and enlarged MAX_COUNT_OF_DEVICES to 5000 but unfortunately this does not solve the issue because only 1000 devices are returned from `registryManager.GetDevicesAsync(maxCountOfDevices);` (DevicesProcessors.cs)

After some googling I found this:
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-bulk-identity-mgmt

And there is written:

> Note
> 
> You can also use the GetDevicesAsync method of the RegistryManager class to fetch a list of your devices. However, this approach has a hard cap of 1000 on the number of device objects that are returned. The expected use case for the GetDevicesAsync method is for development scenarios to aid debugging and is not recommended for production workloads.
> 

In other words we have a challenge here. Anyone able to change this so larger amounts of devices can be loaded?

**Comment by pandazzurro** [comment]
Hi @PieterID, 
You should implement **heartbeat** in your solution.
Like this: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-identity-registry#device-heartbeat.
After that you should store connected devices.

**Comment by PieterID** [comment]
Hi @pandazzurro 

I am not (yet) interested in the info if a device is connected or not I only want to get a list of all registered devices which is at this moment approx 1200 (and growing). I am afraid that we are going to need the bulk features as documented in the link in my issue.

**Comment by pandazzurro** [comment]
Hi @PieterID,
I'm not part of the development team but I started studying technology.
I still do not understand what you need to know about connected devices.

The method you are talking about is not possible to use in production scenarios.
I would implement something like the following:
1) When open device connection, send a message
2) every N minutes indicate device presence in the system
3) Disconnect: notify my cancellation.

The application's backend should only keep a table with connected devices and with the last alert date.
Device management should be your management.

If the purpose is to know the recorded devices, both active and disabled, connected or disconnected ... IMHO would use ExportDevicesAsync method.

I hope I have clarified my previous idea.
At present, the SDK does not implement other methods.

**Comment by PieterID** [comment]
Hi @pandazzurro 

Thanks again for your reply. We are investigating ExportDevicesAsync to see if this brings us a solution. We are not using DeviceExplorer in our production scenario but are using it from time to time to do some short tests and debugging so I think the ExportDevicesAsync is ok for now.

Thanks.
